builtInFunctions/esdtTransferRole.go

```diff
@@ -0,0 +1,145 @@
+package builtInFunctions
+
+import (
+	"bytes"
+	"math/big"
+
+	"github.com/ElrondNetwork/elrond-go-core/core"
+	"github.com/ElrondNetwork/elrond-go-core/core/atomic"
+	"github.com/ElrondNetwork/elrond-go-core/core/check"
+	"github.com/ElrondNetwork/elrond-go-core/marshal"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+const transfer = "transfer"
+
+var transferAddressesKeyPrefix = []byte(core.ElrondProtectedKeyPrefix + transfer + core.ESDTKeyIdentifier)
+
+type esdtTransferAddress struct {
+	*baseEnabled
+	set             bool
+	marshalizer     vmcommon.Marshalizer
+	accounts        vmcommon.AccountsAdapter
+	maxNumAddresses uint32
+}
+
+// NewESDTTransferRoleAddressFunc returns the esdt transfer role address handler built-in function component
+func NewESDTTransferRoleAddressFunc(
+	accounts vmcommon.AccountsAdapter,
+	marshalizer marshal.Marshalizer,
+	activationEpoch uint32,
+	epochNotifier vmcommon.EpochNotifier,
+	maxNumAddresses uint32,
+	set bool,
+) (*esdtTransferAddress, error) {
+	if check.IfNil(marshalizer) {
+		return nil, ErrNilMarshalizer
+	}
+	if check.IfNil(epochNotifier) {
+		return nil, ErrNilEpochHandler
+	}
+	if check.IfNil(accounts) {
+		return nil, ErrNilAccountsAdapter
+	}
+
+	e := &esdtTransferAddress{
+		accounts:        accounts,
+		marshalizer:     marshalizer,
+		maxNumAddresses: maxNumAddresses,
+		set:             set,
+	}
+
+	e.baseEnabled = &baseEnabled{
+		function:        vmcommon.BuiltInFunctionESDTTransferRoleAddAddress,
+		activationEpoch: activationEpoch,
+		flagActivated:   atomic.Flag{},
+	}
+	if !set {
+		e.function = vmcommon.BuiltInFunctionESDTTransferRoleDeleteAddress
+	}
+
+	epochNotifier.RegisterNotifyHandler(e)
+
+	return e, nil
+}
+
+// SetNewGasConfig is called whenever gas cost is changed
+func (e *esdtTransferAddress) SetNewGasConfig(_ *vmcommon.GasCost) {
+}
+
+// ProcessBuiltinFunction resolves ESDT change roles function call
+func (e *esdtTransferAddress) ProcessBuiltinFunction(
+	_, _ vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) (*vmcommon.VMOutput, error) {
+	err := checkBasicESDTArguments(vmInput)
+	if err != nil {
+		return nil, err
+	}
+	if !bytes.Equal(vmInput.CallerAddr, core.ESDTSCAddress) {
+		return nil, ErrAddressIsNotESDTSystemSC
+	}
+
```

Comment:
no check for RecipientAddr ?
```suggestion
if !vmcommon.IsSystemAccountAddress(vmInput.RecipientAddr) {
		return nil, ErrOnlySystemAccountAccepted
	}
```